packages/notebook/src/history.ts

```diff
@@ -0,0 +1,419 @@
+
+  /**
+   * The current editor used by the history manager.
+   */
+  get editor(): CodeEditor.IEditor | null {
+    return this._editor;
+  }
+
+  set editor(value: CodeEditor.IEditor | null) {
+    if (this._editor === value) {
+      return;
+    }
+
+    const prev = this._editor;
+    if (prev) {
+      prev.model.sharedModel.changed.disconnect(this.onTextChange, this);
+    }
+
+    this._editor = value;
+
+    if (value) {
+      value.model.sharedModel.changed.connect(this.onTextChange, this);
+    }
+  }
+
+  /**
+   * The placeholder text that a history session began with.
+   */
+  get placeholder(): string {
+    return this._placeholder;
+  }
+
+  /**
+   * Kernel session number for filtering
+   */
+  get kernelSession(): string {
+    return this._kernelSession;
+  }
+
+  /**
+   * Get whether the console history manager is disposed.
+   */
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  /**
+   * Dispose of the resources held by the console history manager.
+   */
+  dispose(): void {
+    this._isDisposed = true;
+    this._history.length = 0;
+    Signal.clearData(this);
+  }
+
+  /**
+   * Set placeholder and editor. Start session if one is not already started.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   */
+  protected async checkSession(activeCell: Cell): Promise<void> {
+    if (!this._hasSession) {
+      await this._retrieveHistory();
+      this._hasSession = true;
+      this.editor = activeCell.editor;
+      this._placeholder = this._editor?.model.sharedModel.getSource() || '';
+      // Filter the history with the placeholder string.
+      this.setFilter(this._placeholder);
+      this._cursor = this._filtered.length - 1;
+    }
+  }
+
+  /**
+   * Get the previous item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  async back(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    --this._cursor;
+    if (this._cursor < 0) {
+      await this.fetchBatch();
+    }
+    this._cursor = Math.max(0, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  async forward(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    ++this._cursor;
+    this._cursor = Math.min(this._filtered.length - 1, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the promise returned from back or forward
```

Comment:
Let's drop _promise_ as it is the content

```suggestion
   * @param update - the result from back or forward
```